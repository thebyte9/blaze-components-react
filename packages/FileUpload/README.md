
## Description

FileUpload component  is a great draggable area, move one or multiple images to a desired location and "drop" it there using a mouse or similar device.

## Usage

* simple FileUpload

```js

<FileUpload 
    handleDrop={({event, base64, files, canceled}) => {}}>
    // Preview files or any other logic
</FileUpload>

```

* Modal with FileUpload

```js
class ModalWithFileUpload extends Component {
      state = {
        previewImages: [],
        filesToUpload: []
      }

      handleDrop = ({ base64, canceled, files }) => {
        const { previewImages, filesToUpload } = this.state;

        if (canceled) return this.setState({previewImages: []});

        const images = base64.map(src => (
            <img key={uuidv1()} src={src} style={imageStyles}/>
        ));

        this.setState({
          previewImages: [...previewImages, ...images],
          filesToUpload: [...filesToUpload, ...files]
        })
      }

      render() {
        const { state: { previewImages }, handleDrop } = this;
        return (
          <Modal
            isActive
            buttonText="Upload Files"
            title="Add media"
            actions={[['submit', () => {}, 'rounded outline']]}
            upload>
            <FileUpload handleDrop={handleDrop}>
                {previewImages}
            </FileUpload>
          </Modal>
        )
      }
    }

    <ModalWithFileUpload />
```

## API

##### FileUpload can receive a number of `props` as follow:

| NAME   | TYPE | DEFAULT | 
| :---  | :---:  | :---: | 
| handleDrop | Function | () => {} | 
| children | Single or array of nodes | No content |

