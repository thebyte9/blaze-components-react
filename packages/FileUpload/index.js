import React, { useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import uuidv1 from 'uuid/v1';
import Button from '@blaze-react/button';

const FileUpload = ({ children, handleDrop: handleDropProp, ...attr }) => {
  const area = useRef(null);
  const selectFile = useRef(null);

  const handleDragover = event => {
    event.stopPropagation();
    event.preventDefault();
  };

  const getPreview = files =>
    Promise.all(
      files.map(
        file =>
          new Promise((resolve, reject) => {
            if (file.type.includes('image')) {
              const reader = new FileReader();
              reader.readAsDataURL(file);
              reader.onload = e =>
                resolve({
                  id: uuidv1(),
                  name: file.name,
                  base64: e.target.result,
                  type: 'image'
                });
              reader.onerror = () => reject(new DOMException('Error parsing input file.'));
            } else {
              resolve({
                id: uuidv1(),
                name: file.name,
                type: 'file'
              });
            }
          })
      )
    );

  const processFiles = async (event, files) => {
    if (!files || !files.length) return;

    getPreview(files).then(previewFiles => handleDropProp({ event, files, previewFiles }));
  };

  const handleChange = event => {
    event.preventDefault();

    let { target: { files } = {} } = event;

    files = Object.values(files);

    processFiles(event, files);
  };

  useEffect(
    () => {
      const handleDrop = event => {
        event.preventDefault();
        event.stopPropagation();

        let { dataTransfer: { files } = {} } = event;

        files = Object.values(files);

        processFiles(event, files);
      };

      const { current: currentArea } = area;

      currentArea.addEventListener('dragover', handleDragover);
      currentArea.addEventListener('drop', handleDrop);
    },
    [processFiles]
  );

  const handleBrowse = () => {
    const { current: currentSelectFile } = selectFile;
    currentSelectFile.click();
  };

  const handleCancel = event => {
    handleDropProp({ event, canceled: true });
  };

  return (
    <div ref={area} className="upload" {...attr}>
      <i className="material-icons">arrow_upward</i>
      <p>Drag &amp; drop file to upload</p>

      <div className="upload__browse">
        <Button onClick={handleBrowse}>Browse</Button>
        <input type="file" onChange={handleChange} ref={selectFile} style={{ display: 'none' }} />
      </div>

      <div className="upload__text">or</div>

      <Button onClick={handleCancel} modifiers="dark outline">
        Cancel
      </Button>

      {children}
    </div>
  );
};

FileUpload.propTypes = {
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
  handleDrop: PropTypes.func
};

FileUpload.defaultProps = {
  handleDrop: () => {},
  children: 'No content'
};

export default FileUpload;
