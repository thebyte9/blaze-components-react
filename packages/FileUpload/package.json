{
  "name": "@blaze-react/file-upload",
  "version": "0.8.0-alpha.60",
  "description": "FileUpload component",
  "main": "./dist/main.js",
  "scripts": {
    "build": "rm -Rf ./dist && NODE_ENV=production webpack --config webpack.config.js",
    "build-storybook": "build-storybook -c .storybook -o docs-build",
    "lint": "eslint '*/**/*.{ts,tsx}' --ignore-path ../../.eslintignore --fix",
    "test": "jest --passWithNoTests",
    "test:coverage": "jest --coverage --passWithNoTests",
    "prepublishOnly": "yarn build",
    "bundle-analyser": "node ../../node_modules/webpack-bundle-analyzer/lib/bin/analyzer.js ./stats/stats.json"
  },
  "keywords": [],
  "author": "byte9",
  "license": "GPL-3.0",
  "peerDependencies": {
    "@blaze-react/utils": "^0.8.0-alpha.58",
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  },
  "devDependencies": {
    "webpack": "^5.38.1",
    "webpack-cli": "^4.7.0"
  },
  "gitHead": "d9cac6fc28dad1d0794718aac09d292019446698"
}