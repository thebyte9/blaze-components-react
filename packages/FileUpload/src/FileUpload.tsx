import cloneDeep from 'lodash.clonedeep';
import React, { useEffect, useRef, useState } from 'react';
import { nanoid } from 'nanoid';
import Actions from './Actions';
import { DATA_ATTRIBUTS } from './constants';
import { NAME } from './constants';
import DraggableFileUpload from './DraggableFileUpload';
import FileList from './FileList';
interface IFileUploadProps {
  children?: any;
  customPreview?: boolean;
  onChange: (...args: any[]) => void;
  handleDrop?: (...args: any[]) => void;
  handleLibraryClick?: (...args: any[]) => void;
  enableDragAndDrop?: boolean;
  actionText?: any;
  selectOptions: any[];
}
const FileUpload: React.SFC<IFileUploadProps> = ({
  onChange,
  handleDrop: handleDropProp,
  customPreview,
  handleLibraryClick,
  enableDragAndDrop,
  actionText,
  selectOptions,
  ...attr
}) => {
  const [previewImages, setPreviewImages]: any[] = useState([]);
  const [filesToUpload, setFilesToUpload]: any[] = useState([]);
  const area: any = useRef(null);
  const selectFile: any = useRef(null);
  const handleDragover = (event: any): void => {
    event.stopPropagation();
    event.preventDefault();
  };
  useEffect(() => {
    const handler = setTimeout((): void => {
      onChange(filesToUpload);
    }, 500);
    return () => {
      clearTimeout(handler);
    };
  }, [filesToUpload]);
  useEffect(() => {
    const handleDrop = (event: any) => {
      event.preventDefault();
      event.stopPropagation();
      let { dataTransfer: { files = {} } = {} } = event;
      files = Object.values(files);
      processFiles(files);
      onChange(files);
    };
    const { current: currentArea } = area;
    if (enableDragAndDrop) {
      currentArea.addEventListener('dragover', handleDragover);
      currentArea.addEventListener('drop', handleDrop);
      return () => {
        currentArea.removeEventListener('dragover', handleDragover);
        currentArea.removeEventListener('drop', handleDrop);
      };
    }
    return;
  }, [previewImages, filesToUpload]);
  const getPreview = (files: any[]) =>
    Promise.all(
      files.map(
        (file) =>
          new Promise((resolve, reject) => {
            if (file.type && file.type.includes('image')) {
              const reader = new FileReader();
              reader.readAsDataURL(file);
              reader.onload = (e: any) =>
                resolve({
                  data: { ...DATA_ATTRIBUTS },
                  file: {
                    base64: e.target.result,
                    id: file.id,
                    name: file.name,
                    type: 'image',
                  },
                  name: '',
                });
              reader.onerror = () => reject(new DOMException('Error parsing input file.'));
            } else if (file.type && file.type.includes('video')) {
              resolve({
                file: {
                  id: file.id,
                  name: file.name,
                  type: 'video',
                },
                name: '',
              });
            } else {
              resolve({
                data: { ...DATA_ATTRIBUTS },
                file: {
                  id: file.id,
                  name: file.name,
                  type: 'doc',
                },
                name: '',
              });
            }
          }),
      ),
    );
  const processFiles = async (files: any): Promise<any> => {
    if (!files || !files.length) {
      return;
    }
    files = files.map((file: any) => {
      try {
        file.id = nanoid();
      } catch (e) {
        return file;
      }
      return file;
    });
    const previewFiles = await getPreview(files);
    const formatFiles = files.map((file: any) => ({
      data: { ...DATA_ATTRIBUTS },
      file,
    }));
    setFilesToUpload([...filesToUpload, ...formatFiles]);
    setPreviewImages([...previewImages, ...previewFiles]);
    if (handleDropProp) {
      handleDropProp({ previewFiles: [...previewImages, ...previewFiles] });
    }
  };
  const handleChange = (event: any) => {
    event.preventDefault();
    let { target: { files = {} } = {} } = event;
    files = Object.values(files);
    processFiles(files);
    onChange(files);
  };
  const handleBrowse = () => {
    const { current: currentSelectFile } = selectFile;
    currentSelectFile.click();
  };
  const handleCancel = (idToRemove: string): void => {
    const validFiles = (files: any[]) =>
      files.filter(({ file: { id } }: { file: { id: string } }) => id !== idToRemove);
    const fileToUploadUpdated = validFiles(filesToUpload);
    const previewImagesUpdated = validFiles(previewImages);
    setFilesToUpload(fileToUploadUpdated);
    setPreviewImages(previewImagesUpdated);
    onChange(fileToUploadUpdated);
    if (handleDropProp) {
      handleDropProp({
        previewFiles: previewImagesUpdated,
      });
    }
  };
  const copyToOthers = (name: string, index: number) => {
    const filesToUploadCopy = cloneDeep(filesToUpload);
    const previewImagesCopy = cloneDeep(previewImages);

    filesToUploadCopy.forEach((file: any, i: number) => {
      if (name === 'title') {
        file.name = filesToUploadCopy[index].name;
        previewImagesCopy[i].name = filesToUploadCopy[index].name;
      } else {
        file.data[name] = filesToUploadCopy[index].data[name];
        previewImagesCopy[i].data[name] = filesToUploadCopy[index].data[name];
      }
    });
    setFilesToUpload(filesToUploadCopy);
    setPreviewImages(previewImagesCopy);
  };

  const handleInputChange = ({ event }: any) => {
    const {
      target: { id, name, value },
    } = event;
    const filesToUploadCopy = cloneDeep(filesToUpload);
    const previewImagesCopy = cloneDeep(previewImages);
    const index = Number(id.split('-')[0]);
    if (name !== NAME) {
      const _name = name.split('-')[0];
      filesToUploadCopy[index].data[_name] = value;
      previewImagesCopy[index].data[_name] = value;
    } else {
      filesToUploadCopy[index][name] = value;
      previewImagesCopy[index][name] = value;
    }
    setFilesToUpload(filesToUploadCopy);
    setPreviewImages(previewImagesCopy);
  };
  const handleSelectChange = ({ event }: any, index: number) => {
    const {
      target: { value },
    } = event;
    const filesToUploadCopy = cloneDeep(filesToUpload);
    filesToUploadCopy[index].storeKey = value;
    setFilesToUpload(filesToUploadCopy);
  };
  return (
    <>
      {enableDragAndDrop ? (
        <DraggableFileUpload
          previewImages={previewImages}
          handleCancel={handleCancel}
          customPreview={customPreview}
          area={area}
          handleInputChange={handleInputChange}
          handleSelectChange={handleSelectChange}
          selectOptions={selectOptions}
          copyToOthers={copyToOthers}
          {...attr}
        >
          <Actions
            actionText={actionText}
            handleLibraryClick={handleLibraryClick}
            handleBrowse={handleBrowse}
            handleChange={handleChange}
            selectFile={selectFile}
          />
        </DraggableFileUpload>
      ) : (
        <>
          <Actions
            actionText={actionText}
            handleLibraryClick={handleLibraryClick}
            handleBrowse={handleBrowse}
            handleChange={handleChange}
            selectFile={selectFile}
          />
          {!customPreview && (
            <FileList
              previewImages={previewImages}
              handleCancel={handleCancel}
              handleInputChange={handleInputChange}
              handleSelectChange={handleSelectChange}
              selectOptions={selectOptions}
              copyToOthers={copyToOthers}
            />
          )}
        </>
      )}
    </>
  );
};
FileUpload.defaultProps = {
  actionText: 'Add Files',
  children: 'No content',
  customPreview: false,
  enableDragAndDrop: true,
  handleDrop: () => void 0,
  onChange: () => void 0,
};
export default FileUpload;
