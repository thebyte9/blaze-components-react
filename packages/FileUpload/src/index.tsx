import Button from "@blaze-react/button";
import React, { useEffect, useRef } from "react";
import uuidv1 from "uuid/v1";
interface IFileUploadProps {
  handleDrop: (...args: any[]) => any;
  children?: any;
}
const FileUpload: React.SFC<IFileUploadProps> = ({
  children,
  handleDrop: handleDropProp,
  ...attr
}) => {
  const area: any = useRef(null);
  const selectFile: any = useRef(null);
  const handleDragover = (event: any): void => {
    event.stopPropagation();
    event.preventDefault();
  };
  const getPreview = (files: any[]) =>
    Promise.all(
      files.map(
        file =>
          new Promise((resolve, reject) => {
            if (file.type.includes("image")) {
              const reader = new FileReader();
              reader.readAsDataURL(file);
              reader.onload = (e: any) =>
                resolve({
                  base64: e.target.result,
                  id: uuidv1(),
                  name: file.name,
                  type: "image"
                });
              reader.onerror = () =>
                reject(new DOMException("Error parsing input file."));
            } else {
              resolve({
                id: uuidv1(),
                name: file.name,
                type: "file"
              });
            }
          })
      )
    );
  const processFiles = async (event: any, files: any): Promise<any> => {
    if (!files || !files.length) {
      return;
    }
    getPreview(files).then(
      (previewFiles: any): void => {
        handleDropProp({ event, files, previewFiles });
      }
    );
  };
  const handleChange = (event: any) => {
    event.preventDefault();
    let { target: { files = {} } = {} } = event;
    files = Object.values(files);
    processFiles(event, files);
  };
  useEffect(() => {
    const handleDrop = (event: any) => {
      event.preventDefault();
      event.stopPropagation();
      let { dataTransfer: { files = {} } = {} } = event;
      files = Object.values(files);
      processFiles(event, files);
    };
    const { current: currentArea } = area;
    currentArea.addEventListener("dragover", handleDragover);
    currentArea.addEventListener("drop", handleDrop);
  }, [processFiles]);
  const handleBrowse = () => {
    const { current: currentSelectFile } = selectFile;
    currentSelectFile.click();
  };
  const handleCancel = (event: any) => {
    handleDropProp({ event, canceled: true });
  };
  return (
    <div ref={area} className="upload" {...attr}>
      <i className="material-icons">arrow_upward</i>
      <p>Drag &amp; drop file to upload</p>

      <div className="upload__browse">
        <Button onClick={handleBrowse}>Browse</Button>
        <input
          type="file"
          onChange={handleChange}
          ref={selectFile}
          style={{ display: "none" }}
        />
      </div>

      <div className="upload__text">or</div>

      <Button
        onClick={handleCancel}
        modifiers={[
          Button.availableModifiers.outline,
          Button.availableModifiers.dark
        ]}
      >
        Cancel
      </Button>

      {children}
    </div>
  );
};
FileUpload.defaultProps = {
  children: "No content",
  handleDrop: (): void => {
    return;
  }
};
export default FileUpload;
