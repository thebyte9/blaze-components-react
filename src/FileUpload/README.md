
## Description

FileUpload component  is a great draggable area, move one or multiple images to a desired location and "drop" it there using a mouse or similar device.

## Usage

* simple FileUpload

```js

<FileUpload 
    handleDrop={({event, base64, files, canceled}) => {}}>
    // Preview files or any other logic
</FileUpload>

```

* Modal with FileUpload

```js
class ModalWithFileUpload extends Component {
  state = {
    preview: [],
    filesToUpload: []
  }

  removeFile = (id) => {
    const preview = this.state.preview.filter(file => file.id !== id);
    this.setState({ preview: preview });
  }

  previewImage = (base64, name, id) => (
    <div key={id}>
      <img src={base64} alt="preview image" style={imageStyles}/>
      <span>{name}</span>
      <i onClick={() => this.removeFile(id)} className="material-icons">clear</i>
    </div>
  )

  previewFile = (name, id) => (
    <div key={id}>
      <i className="material-icons">attach_file</i>
      <span>{name}</span>
      <i onClick={() => this.removeFile(id)} className="material-icons">clear</i>
    </div>
  )

  handleDrop = ({ canceled, files, previewFiles }) => {
    const { preview, filesToUpload } = this.state;

    if (canceled) return this.setState({preview: []});

    this.setState({
      preview: [...preview, ...previewFiles],
      filesToUpload: [...filesToUpload, ...files]
    })
  }

  render() {
    const { handleDrop } = this;
    const previewFiles = this.state.preview.map(({name, base64, type, id}) => {
      if (type === 'image') return this.previewImage(base64, name, id)
      return this.previewFile(name, id)
    });
    return (
      <Modal
      isActive
      buttonText="Upload Files"
      title="Add media"
      actions={[['submit', () => {}, 'rounded outline']]}
      upload>
        <FileUpload handleDrop={handleDrop} style={styles}>
            {previewFiles}
        </FileUpload>
      </Modal>
    )
  }
}

    <ModalWithFileUpload />
```

## API

##### FileUpload can receive a number of `props` as follow:

| NAME   | TYPE | DEFAULT | 
| :---  | :---:  | :---: | 
| handleDrop | Function | () => {} | 
| children | Single or array of nodes | No content |

