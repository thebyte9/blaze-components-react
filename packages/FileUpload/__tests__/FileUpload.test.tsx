import React from 'react';
import expect from 'expect';
import { shallow, mount } from 'enzyme';
import FileUpload from '../src';

declare global {
  interface Window {
    FileReader: any;
  }
}

const FileUploadComponent = <FileUpload handleDrop={() => { }}>Drag and drop here</FileUpload>;

describe('FileUpload component', () => {
  test('should be defined and renders correctly (snapshot)', () => {
    const wrapper = shallow(FileUploadComponent);

    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  test('should cancel operation', () => {
    const wrapper = mount(FileUploadComponent);

    wrapper
      .find('Button')
      .at(1)
      .simulate('click');
  });

  test('should browse file', async done => {
    const wrapper = mount(FileUploadComponent);

    wrapper
      .find('Button')
      .at(0)
      .simulate('click');

    const typeImage = new Blob(['file contents'], { type: 'image/png' });
    const typeFile = new Blob(['file contents'], { type: 'application/pdf' });

    const readAsDataURL = jest.fn();
    const onload = jest.fn();
    const onerror = jest.fn();

    const dummyFileReader = {
      readAsDataURL,
      onload,
      onerror
    };

    window.FileReader = jest.fn(() => dummyFileReader);

    wrapper
      .find('input')
      .at(0)
      .simulate('change', { target: { files: [typeImage] } });
    wrapper
      .find('input')
      .at(0)
      .simulate('change', { target: { files: [typeFile] } });
    wrapper
      .find('input')
      .at(0)
      .simulate('change', { target: { files: [] } });
    done();
  });

  test('should drop files', () => {
    const wrapper = mount(FileUploadComponent);

    const domNode = wrapper.getDOMNode();

    const dragover = new Event('dragover');

    const drop: CustomEvent & { dataTransfer?: any } = new CustomEvent('drop', { bubbles: true, cancelable: true });

    const file = {
      name: 'test.jpg',
      type: 'image/jpg',
    } as File;

    const fileList = {
      length: 1,
      item: () => null,
      0: file,
    }


    drop.dataTransfer = {
      files: fileList
    };

    const dropWithoutItems: CustomEvent & { dataTransfer?: any } = new CustomEvent('drop', { bubbles: true, cancelable: true });
    dropWithoutItems.dataTransfer = { files: fileList };

    domNode.dispatchEvent(dragover);
    domNode.dispatchEvent(drop);
    domNode.dispatchEvent(dropWithoutItems);
  });
});
