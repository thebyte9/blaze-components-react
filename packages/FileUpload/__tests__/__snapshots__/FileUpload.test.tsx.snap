// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUpload component should be defined and renders correctly (snapshot) 1`] = `
<DocumentFragment>
  <div
    class="upload"
  >
    <div
      class="upload__drag-drop"
    >
      <div
        class="upload__icon"
      >
        <i
          class="material-icons"
        >
          arrow_upward
        </i>
      </div>
      <p>
        Drag & drop file to upload
      </p>
    </div>
    <div
      class="upload__browse"
    >
      <div
        class="upload__text"
      >
        or
      </div>
      <button
        class="button"
      >
        Browse
      </button>
      <input
        style="display: none;"
        type="file"
      />
    </div>
  </div>
</DocumentFragment>
`;
